server/src/main/java/org/elasticsearch/index/codec/vectors/DefaultIVFVectorsReader.java

```diff
@@ -0,0 +1,420 @@
+        OptimizedScalarQuantizer.QuantizationResult queryParams = scalarQuantizer.scalarQuantize(
+            targetScratch,
+            quantized,
+            (byte) 4,
+            globalCentroid
+        );
+        return new CentroidQueryScorer() {
+            int currentCentroid = -1;
+            private final byte[] quantizedCentroid = new byte[fieldInfo.getVectorDimension()];
+            private final float[] centroid = new float[fieldInfo.getVectorDimension()];
+            private final float[] centroidCorrectiveValues = new float[3];
+            private int quantizedCentroidComponentSum;
+            private final long centroidByteSize = fieldInfo.getVectorDimension() + 3 * Float.BYTES + Short.BYTES;
+
+            @Override
+            public int size() {
+                return numCentroids;
+            }
+
+            @Override
+            public float[] centroid(int centroidOrdinal) throws IOException {
+                readQuantizedCentroid(centroidOrdinal);
+                return centroid;
+            }
+
+            private void readQuantizedCentroid(int centroidOrdinal) throws IOException {
+                if (centroidOrdinal == currentCentroid) {
+                    return;
+                }
+                centroids.seek(centroidOrdinal * centroidByteSize);
+                quantizedCentroidComponentSum = readQuantizedValue(centroids, quantizedCentroid, centroidCorrectiveValues);
+                centroids.seek(numCentroids * centroidByteSize + (long) Float.BYTES * quantizedCentroid.length * centroidOrdinal);
+                centroids.readFloats(centroid, 0, centroid.length);
+                currentCentroid = centroidOrdinal;
+            }
+
+            @Override
+            public float score(int centroidOrdinal) throws IOException {
+                readQuantizedCentroid(centroidOrdinal);
+                return int4QuantizedScore(
+                    quantized,
+                    queryParams,
+                    fieldInfo.getVectorDimension(),
+                    quantizedCentroid,
+                    centroidCorrectiveValues,
+                    quantizedCentroidComponentSum,
+                    globalCentroidDp,
+                    fieldInfo.getVectorSimilarityFunction()
+                );
+            }
+        };
+    }
+
+    @Override
+    protected FloatVectorValues getCentroids(IndexInput indexInput, int numCentroids, FieldInfo info) {
+        FieldEntry entry = fields.get(info.number);
+        if (entry == null) {
+            return null;
+        }
+        return new OffHeapCentroidFloatVectorValues(numCentroids, indexInput, info.getVectorDimension());
+    }
+
+    @Override
+    NeighborQueue scorePostingLists(FieldInfo fieldInfo, KnnCollector knnCollector, CentroidQueryScorer centroidQueryScorer, int nProbe)
+        throws IOException {
+        NeighborQueue neighborQueue = new NeighborQueue(centroidQueryScorer.size(), true);
```

Comment:
Are you supposed to have MIN heap of only `nprobe` size?

Reply:
For this, we need to be able to explore past nProbe in the case of filtered search. `nProbe` has no current role. 

I think we will end up with a double queue system eventually, but given its all flat, we just score everything. This needs to be improved, for sure.